src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/ReflectionJsonTypeInfoOfT.cs

```diff
@@ -66,7 +71,11 @@ private void AddPropertiesAndParametersUsingReflection()
             // Walk through the inheritance hierarchy, starting from the most derived type upward.
             for (Type? currentType = Type; currentType != null; currentType = currentType.BaseType)
             {
-                PropertyInfo[] properties = currentType.GetProperties(BindingFlags);
+                var bindingFlags = BindingFlagsDeclaredOnly;
```

Comment:
Do we really need to use different binding flags for interface types? The reason the bug exists is because `currentType.BaseType` always returns null for interface types so the for loop will only iterate once and not consider any base interfaces. What I would recommend doing instead is tweaking the loop so that it iterates over `currentType.GetInterfaces()` in cases where `currentType` is an interface specifically.

Reply:
Do you think that is faster? What is if some interfaces define the same properties?

Reply:
I've not yet tested the change. I need to check how to compile files of the .net runtime repo. I only wanted to know at first if fixing it like this would be okay.

Reply:
I agree with @eiriktsarpalis here, removing DeclaredOnly might end up showing same properties couple of times.

I'm curious if we could simplify this code or repeat what other serializers do (i.e. what does XmlSerializer do)

Reply:
I did not simply remove it. I only removed it when the type is an interface

Reply:
I believe the point is that we needn't use different binding flags when the type is an interface. The logic should follow classes in explicitly walking the type hierarchy and resolve any ambiguity following the same logic.

Reply:
> Do you think that is faster? What is if some interfaces define the same properties?

It's not an issue of performance. The CacheMember method applies its own logic to resolve conflicts and for the sake of consistency the same logic should be followed in interface.

That being said, one thing that needs to be taken into account is diamond ambiguity which is not possible in class hierarchies.

Reply:
IMO we should see what other serializers are doing and match this behavior